File: tests/manual/signal_processing/agc_test.py
import numpy as np
import matplotlib.pyplot as plt
import sys

# This script plots the results of the AGC (Automatic Gain Control) filter
# applied to a test signal. The signal is generated as CSV by agc_test.cpp
# accepted on stdin.
#
# The test signal has varying amplitude to demonstrate the AGC's ability to
# normalize the signal level.

# Read metadata lines
metadata = {}
for line_num in range(4):
    line = sys.stdin.readline()
    if line.startswith('#'):
        key_value = line[1:].strip().split('=')
        if len(key_value) == 2:
            key = key_value[0].strip()
            value = float(key_value[1].strip())
            metadata[key] = value

sampling_freq = metadata['SAMPLING_FREQ']
target_rms = metadata['TARGET_RMS']
attack_time_sec = metadata['ATTACK_TIME_SEC']
release_time_sec = metadata['RELEASE_TIME_SEC']

# Read CSV data
labels = np.loadtxt(sys.stdin, delimiter=',', max_rows=1, dtype=str)
data = np.loadtxt(sys.stdin, delimiter=',')

# Extract columns
sample_idx = data[:, 0]
input_signal = data[:, 1]
output_signal = data[:, 2]
gain = data[:, 3]
current_rms = data[:, 4]

# Create time axis in seconds
time = sample_idx / sampling_freq

# Calculate RMS of output signal in a sliding window for verification
window_size = int(sampling_freq / 10)  # 100ms window
output_rms_windowed = np.zeros_like(output_signal)
for i in range(len(output_signal)):
    start_idx = max(0, i - window_size)
    end_idx = i + 1
    output_rms_windowed[i] = np.sqrt(
        np.mean(output_signal[start_idx:end_idx]**2))

# Create subplots
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(14, 10))

# Plot 1: Input vs Output signals
ax1.plot(time, input_signal, label='Input', alpha=0.7, linewidth=0.8)
ax1.plot(time, output_signal, label='Output (AGC)', alpha=0.7, linewidth=0.8)
ax1.set_xlabel('Time (s)')
ax1.set_ylabel('Amplitude')
ax1.set_title('AGC Filter - Input vs Output Signals')
ax1.legend()
ax1.grid(alpha=0.3)

# Plot 2: Gain over time
ax2.plot(time, gain, label='Gain', color='green', linewidth=1.5)
ax2.set_xlabel('Time (s)')
ax2.set_ylabel('Gain')
ax2.set_title('AGC Gain Adjustment Over Time')
ax2.legend()
ax2.grid(alpha=0.3)

# Plot 3: RMS tracking
ax3.plot(time,
         current_rms,
         label='Current RMS (estimated)',
         color='blue',
         linewidth=1.5)
ax3.plot(time,
         output_rms_windowed,
         label='Output RMS (windowed)',
         color='orange',
         alpha=0.7,
         linewidth=1.0,
         linestyle='--')
ax3.axhline(y=target_rms,
            color='red',
            linestyle='--',
            linewidth=2,
            alpha=0.7,
            label=f'Target RMS: {target_rms}')
ax3.set_xlabel('Time (s)')
ax3.set_ylabel('RMS')
ax3.set_title('RMS Tracking (Attack: {:.1f}s, Release: {:.1f}s)'.format(
    attack_time_sec, release_time_sec))
ax3.legend()
ax3.grid(alpha=0.3)

# Plot 4: Input and Output envelopes (absolute values with smoothing)
input_envelope_window = int(sampling_freq / 50)  # 20ms window
input_envelope = np.convolve(np.abs(input_signal),
                             np.ones(input_envelope_window) /
                             input_envelope_window,
                             mode='same')
output_envelope = np.convolve(np.abs(output_signal),
                              np.ones(input_envelope_window) /
                              input_envelope_window,
                              mode='same')

ax4.plot(time, input_envelope, label='Input Envelope', color='blue', alpha=0.7)
ax4.plot(time,
         output_envelope,
         label='Output Envelope',
         color='orange',
         alpha=0.7)
# Target RMS line (multiplied by sqrt(2) to get approximate peak for sine wave)
target_peak = target_rms * np.sqrt(2)
ax4.axhline(y=target_peak,
            color='red',
            linestyle='--',
            linewidth=2,
            alpha=0.7,
            label=f'Target Peak: {target_peak:.3f}')
ax4.set_xlabel('Time (s)')
ax4.set_ylabel('Amplitude')
ax4.set_title('Signal Envelopes')
ax4.legend()
ax4.grid(alpha=0.3)

plt.tight_layout()
plt.show()

File: tests/manual/signal_processing/agc_test.cpp
#include <signal_processing/pipeline.hpp>
#include <signal_processing/filters.hpp>

#include <chrono>
#include <cstdint>
#include <vector>
#include <numbers>
#include <cmath>
#include <iostream>

int main() {
    constexpr float SAMPLING_FREQ = 1000.f;
    constexpr size_t TEST_SAMPLES = 5000;

    // Generate a signal with varying amplitude
    // First half: amplitude 0.2, Second half: amplitude 1.0
    std::vector<float> input_signal(TEST_SAMPLES);
    constexpr float freq = 10.f; // 10 Hz sine wave
    for (size_t i = 0; i < TEST_SAMPLES; ++i) {
        float t = static_cast<float>(i) / SAMPLING_FREQ;
        float amplitude = (i < TEST_SAMPLES / 2) ? 0.2f : 1.0f;
        input_signal[i] = amplitude * std::sin(2.0f * std::numbers::pi_v<float> * freq * t);
    }

    // Apply AGC to normalize to target RMS
    using namespace sp::pipe;

    constexpr float TARGET_RMS = 0.5f;
    const auto ATTACK_TIME = std::chrono::milliseconds(100);
    const auto RELEASE_TIME = std::chrono::milliseconds(100);

    auto agc_pipeline = make_source(input_signal, SAMPLING_FREQ)
        | agc(TARGET_RMS, ATTACK_TIME, RELEASE_TIME);

    // Output metadata
    std::cout << "# SAMPLING_FREQ=" << SAMPLING_FREQ << "\n";
    std::cout << "# TARGET_RMS=" << TARGET_RMS << "\n";
    const auto attack_seconds = std::chrono::duration_cast<std::chrono::duration<float>>(ATTACK_TIME).count();
    const auto release_seconds = std::chrono::duration_cast<std::chrono::duration<float>>(RELEASE_TIME).count();
    std::cout << "# ATTACK_TIME_SEC=" << attack_seconds << "\n";
    std::cout << "# RELEASE_TIME_SEC=" << release_seconds << "\n";

    // Output CSV data
    std::cout << "Sample,Input,Output,Gain,CurrentRMS\n";
    for (size_t i = 0; i < TEST_SAMPLES; ++i) {
        float in = input_signal[i];

        // Get the filter for inspection
        auto &filter = agc_pipeline.get_filter();

        float out = agc_pipeline.next();
        float gain = filter.get_current_gain();
        float current_rms = filter.get_current_rms();

        std::cout << i << ","
                  << in << ","
                  << out << ","
                  << gain << ","
                  << current_rms << "\n";
    }

    return 0;
}
